### DISCUSSION TITLE
How to handle AMQP messages

### DISCUSSION THREAD

**Discussion by aeb-dev** [rabbitmq/rabbitmq-stream-dotnet-client#238] [Q&A] [answered]
Hi,

I have a fanout exchange that publishes to both classic queue and stream. When I consume from stream I get extra values inside `Message.Data.Contents` appended to my original payload. I believe it is related to AMQP. The extra data looks similar to the following:

```
x-routing-key�somevalue�
x-exchange�somevalue
```

I could not copy all the extra data because browser does not let me paste it has weird characters

I have seen [this issue](https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/issues/58). I got confused whether should I parse this myself or the stream client should handle it?

**Comment by Gsantomaggio** [comment]
Which RabbitMQ and Client library are you using?

**Comment by aeb-dev** [comment]
RabbitMQ 3.11.5
Stream Client 1.2.0

**Comment by Gsantomaggio** [comment]
I can't reproduce the issue:
![Screenshot 2023-02-21 at 12 08 45](https://user-images.githubusercontent.com/386987/220331241-bbc356d4-5c9c-4f89-8868-27b6dcdb000f.png)

Then I publish a message from the management UI:
![Screenshot 2023-02-21 at 12 20 48](https://user-images.githubusercontent.com/386987/220331563-774154ee-e97b-4bda-98de-b86977c03f70.png)


The amqp values a correctly stored inside the annotation field
![Screenshot 2023-02-21 at 12 17 37](https://user-images.githubusercontent.com/386987/220331656-7a133aa7-2d09-4b5d-8500-66acdfdb1e69.png)



>I could not copy all the extra data because browser does not let me paste it has weird characters

Not clear to me here from where you are trying to copy the values

**Comment by Gsantomaggio** [comment]
Can you please provide a repo where I can reproduce the issue?

**Comment by aeb-dev** [comment]
> Not clear to me here from where you are trying to copy the values

I write the contents of `Message.Data.Contents` to a file. That file contains amqp values. I was not expecting amqp values inside `Contents`

**Comment by Gsantomaggio** [comment]
@aeb-dev, if you don't provide more information I am afraid that we can't help.

- How do you save the message into the files?
- is it systematic? 
- Can you please provide the code to reproduce the issue, we are not able to reproduce it

**Comment by aeb-dev** [comment]
I will try to provide a reproducible example

**Comment by aeb-dev** [comment]
@Gsantomaggio  I want to make sure that I understand it right. Using your example, when I deserialize `Message.Data.Contents` to string I should only get "aaaaaaaa" without any amqp values right?

**Reply by Gsantomaggio** [reply]
No, in my example you should see:
```
My message 5
```
Inside `Message.Data.Contents`

and all the routing information in the `Annotation` field, like:
```
x-routing-key = aaaaaaa
etc... 
```

**Reply by aeb-dev** [reply]
Sorry my bad, that is what I meant

**Comment by aeb-dev** [comment]
This is probably related to us. I will report my findings when I find the problem

**Comment by aeb-dev** [comment]
Sorry for late reply. You can use this repo to reproduce: https://github.com/aeb-dev/rabbitmq-stream-bug

There are 3 variables you should configure to get the result.
`MESSAGE_COUNT`: number of messages to produce & consume. You can play with this if you like...
`consumer`: whether to run consumer or producer. Initially, you should start with `consumer` set to `false` to create data. Then you can consume the same data over and over again.
`working`: When this is set `true` the file will have valid json(Ignore the last comma). If it is set to false it will run the bugged code. You might need to run bugged code multiple times to get the wrong output

**Reply by Gsantomaggio** [reply]
The example seems complex; what if you write directly to the file in the consumer handler? here:
https://github.com/aeb-dev/rabbitmq-stream-bug/blob/main/Program.cs#L131

**Reply by aeb-dev** [reply]
It works normally when you do it like that

**Comment by jonnepmyra** [comment, marked as answer]
Don't think you're supposed to store a reference to a `Message` outside the scope of the `Messagehandler`

`Message.Data` is a `ReadOnlySequence<byte>` and in general, it is not safe to store references to ReadOnlySequence instances for later use.

It's underlying memory regions could for example be re-used by the runtime at any time. I guess that's why you end up with corrupted data (json output) in the `Bugged()` scenario?

**Reply by lukebakken** [reply]
@aeb-dev ^^^^ this is why `Bugged()` does not work correctly. [This code](https://github.com/aeb-dev/rabbitmq-stream-bug/blob/main/Program.cs#L131-L132) works correctly because it copies the data.

**Reply by aeb-dev** [reply]
I was actually suspecting that but although I am not sure about the alternative (Memory maybe ?) but creating another copy of `byte[]` does not make sense?

**Reply by lukebakken** [reply]
> but creating another copy of byte[] does not make sense?

Why not? When data is read from the socket it has to go _somewhere_, so we rent memory from the `ArrayPool<byte>.Shared` object. The `ReadOnlySequence<byte>` is a "window" into that rented memory, and once the scope of the callback is gone the memory isn't yours anymore. You must copy the contents somehow - either into your own memory or parse it into an object.

**Reply by aeb-dev** [reply]
Can't it be made in such a way that it would create one less copy. Although I am not sure about cost but depending on the size it might be better:

Currently: socket buffer -> stream client creates array pool -> I create another copy
Suggestion: socket buffer -> stream client creates memory -> I use the same memory

**Reply by lukebakken** [reply]
> Currently: socket buffer -> stream client creates array pool -> I create another copy

This isn't how it currently works. Please read [the documentation](https://learn.microsoft.com/en-us/dotnet/api/system.buffers.arraypool-1?view=net-7.0).
